docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+
+```bash
+datastore/
+└── motd
+    ├── v1
+    │   └── motd.json
+    └── v2
+        └── motd.json
+```
+Any data written to directories created during a `set` transaction is moved into the datastore.
+Despite only one version of the data being provided, the `apiserver` will populate all versions supported by the settings extension at this time by requesting that the settings extension perform all necessary migrations.
+Failure to populate a version listed as "supported" will result in a failed transaction.
+See the next section about migrations for more details.
+
+### Extension Versioning and Migrations
+Bottlerocket's settings data model changes frequently: sometimes to support new features, or sometimes to correct mistaken model shapes.
+In order to support changes to the settings model, settings must be capable of exposing versioning information about the settings extension, as well as providing faculties for migrating settings data to new versions.
+Our versioning scheme will take heavy influence from Kubernetes' scheme for [versioning CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).
+Settings extension versioning information is surfaced through the extension config file:
+
+```toml
+# hostname.toml -- setting extension config file for settings.hostname
+[extension]
+supported-versions = [
+     "v1",
+     "v2"
+]
+default-version = "v2"
+```
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version` and should only take place when a user makes a request using the CLI -- all automated process *should* refer to an explicit version.
+During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
+These migrations always begin at the `canonical` version, and settings extensions are free to perform the migrations however they wish; however, Bottlerocket's Settings SDK will provide utilities to help settings developers support a series of linear setting upgrades and downgrades.
+Read on for more information about the Settings SDK
+
+All migrations are performed and stored at setting-write-time because Bottlerocket rollbacks may result in running an OS with a settings extension which does not support the version at which a given setting is stored.
+If Bottlerocket boots and finds that a setting is stored at a value which is not supported by it's installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent dataloss in the case that the image can eventually update to a new image which supports the stored setting version.
```

Comment:
```suggestion
If Bottlerocket boots and finds that a setting is stored at a value which is not supported by its installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent data loss in the case that the image can eventually update to a new image which supports the stored setting version.
```